internal/hud/server/controller.go

```diff
@@ -54,26 +62,67 @@ func (s *HeadsUpServerController) OnChange(ctx context.Context, st store.RStore)
 		return
 	}
 
-	l, err := net.Listen("tcp", fmt.Sprintf("%s:%d", string(s.host), int(s.port)))
+	ctx, cancel := context.WithCancel(ctx)
+	s.shutdown = cancel
+
+	err := s.serve(ctx, st)
 	if err != nil {
 		st.Dispatch(
 			store.NewErrorAction(
-				fmt.Errorf("Tilt cannot start because you already have another process on port %d\n"+
-					"If you want to run multiple Tilt instances simultaneously,\n"+
-					"use the --port flag or TILT_PORT env variable to set a custom port\nOriginal error: %v",
-					s.port, err)))
+				fmt.Errorf("Cannot start the tilt-apiserver: %v", err)))
 		return
 	}
+}
 
-	httpServer := &http.Server{
-		Handler: http.DefaultServeMux,
+// Merge the APIServer and the Tilt Web server into a single handler,
+// and attach them both to the public listener.
+//
+// TODO(nick): We could move this to SetUp if SetUp had error-handling.
+func (s *HeadsUpServerController) serve(ctx context.Context, st store.RStore) error {
+	stopCh := ctx.Done()
+	o := s.apiServerOptions
+	config, err := o.Config()
+	if err != nil {
+		return err
 	}
-	http.Handle("/", s.hudServer.Router())
 
-	go func() {
-		<-ctx.Done()
-		_ = httpServer.Shutdown(context.Background())
-	}()
+	server, err := config.Complete().New()
+	if err != nil {
+		return err
+	}
+
+	server.GenericAPIServer.AddPostStartHookOrDie("start-tilt-server-informers", func(context genericapiserver.PostStartHookContext) error {
```

Comment:
Not hugely important (I'm assuming this can't practically fail) but could use `AddPostStartHook` and propagate error to avoid surprise panic

Reply:
ya, you're right on both counts - it really only fails if you've registered a conflicting hook, and it also seems safer to use the error-returning function. done!